packages/local_auth/local_auth_android/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java

```diff
@@ -92,176 +90,112 @@ public static void registerWith(PluginRegistry.Registrar registrar) {
+          new AuthResultWrapper.Builder().setValue(AuthResult.ERROR_NOT_FRAGMENT_ACTIVITY).build());
       return;
     }
 
     if (!isDeviceSupported()) {
-      authInProgress.set(false);
-      result.error("NotAvailable", "Required security features not enabled", null);
+      result.success(
+          new AuthResultWrapper.Builder().setValue(AuthResult.ERROR_NOT_AVAILABLE).build());
       return;
     }
 
     authInProgress.set(true);
     AuthCompletionHandler completionHandler = createAuthCompletionHandler(result);
 
-    boolean isBiometricOnly = call.argument("biometricOnly");
-    boolean allowCredentials = !isBiometricOnly && canAuthenticateWithDeviceCredential();
+    boolean allowCredentials = !options.getBiometricOnly() && canAuthenticateWithDeviceCredential();
 
-    sendAuthenticationRequest(call, completionHandler, allowCredentials);
-    return;
+    sendAuthenticationRequest(options, strings, allowCredentials, completionHandler);
   }
 
   @VisibleForTesting
-  public AuthCompletionHandler createAuthCompletionHandler(final Result result) {
-    return new AuthCompletionHandler() {
-      @Override
-      public void onSuccess() {
-        authenticateSuccess(result);
-      }
-
-      @Override
-      public void onFailure() {
-        authenticateFail(result);
-      }
-
-      @Override
-      public void onError(String code, String error) {
-        if (authInProgress.compareAndSet(true, false)) {
-          result.error(code, error, null);
-        }
-      }
-    };
+  public @NonNull AuthCompletionHandler createAuthCompletionHandler(
+      @NonNull final Result<AuthResultWrapper> result) {
+    return authResult -> onAuthenticationCompleted(result, authResult);
   }
 
   @VisibleForTesting
   public void sendAuthenticationRequest(
-      MethodCall call, AuthCompletionHandler completionHandler, boolean allowCredentials) {
+      @NonNull AuthOptions options,
+      @NonNull AuthStrings strings,
+      boolean allowCredentials,
+      @NonNull AuthCompletionHandler completionHandler) {
     authHelper =
         new AuthenticationHelper(
-            lifecycle, (FragmentActivity) activity, call, completionHandler, allowCredentials);
+            lifecycle,
+            (FragmentActivity) activity,
+            options,
+            strings,
+            completionHandler,
+            allowCredentials);
 
     authHelper.authenticate();
   }
 
-  private void authenticateSuccess(Result result) {
+  void onAuthenticationCompleted(Result<AuthResultWrapper> result, AuthResult value) {
     if (authInProgress.compareAndSet(true, false)) {
-      result.success(true);
+      result.success(new AuthResultWrapper.Builder().setValue(value).build());
     }
   }
 
-  private void authenticateFail(Result result) {
-    if (authInProgress.compareAndSet(true, false)) {
-      result.success(false);
-    }
-  }
-
-  /*
-   * Stops the authentication if in progress.
-   */
-  private void stopAuthentication(Result result) {
```

Comment:
Almost all of the "removed" code in this removed-code diff block is actually just moved up inline into the new Pigeon API implementation functions above.